File: helper_funcs/test_cartpole_open_loop.m
clearvars; clc;

plant = cartpole();
plant.animate = true;
plant.animationFig = figure('WindowState', 'maximized');
plant.pauseTime = 0.005; % [s] pause time between animation frames

h = 0.025;             % [s] sampling interval
t = 0:h:20;            % [s] time vector
Nsim = numel(t) - 1;   % number of time samples
x0 = [0; 0; 0; 0];    % [m; m/s; rad; rad/s] initial state 

% apply some control action F: during the first quarter of time F=0.2 N.
% Aftewards, F=-0.2 N;
Usim = 0.2*ones(1, Nsim);
Usim(round(0.18*Nsim):round(0.58*Nsim)) = -0.2;

Xsim = NaN(4, Nsim + 1);
Xsim(:, 1) = x0;
plant.x = x0;

for k=1:Nsim
    Xsim(:, k+1) = plant.simulate(Usim(:,k), h);
end

%% plot
figure; 
plot(t, Xsim(4,:)); grid on;
xlabel("Time (s)");
ylabel("d\theta/dt");
title("Open-Loop simulation of the cart-pole system");

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   Created by Nikolce Murgovski, 2025-12.

File: helper_funcs/test_cartpole_open_loop.py
from cartpole import CartPole
import numpy as np
import matplotlib.pyplot as plt

cp = CartPole(animate=True) # instantiate the cart-pole simulator
cp.pauseTime= 0.01          # pause between plot frames
cp.set_animation_figure()   # creates a figure (if you dont run this line, the simulator will Not produce a plot)
x0 = np.array([0.0, 0.0, 0.0, 0.0] )    # initial state
cp.x[:] = x0  
dt = 0.025          # sampling time
T = 20              # simulation time
t = np.arange(0,T,dt)   # [s] time vector
Nsim = t.shape[0]

# test control
Usim = 0.2*np.ones(Nsim) # push to the right for some time
Usim[144:464 ] = -0.2       # then push to the left

# simulation loop
Xsim = np.ones((Nsim,4))*np.inf
Xsim[0,:] = x0
for k in range(Nsim):
    cp.simulate(F=Usim[k], dt=dt, disturbance=0.0,title=f'Time step {k}/{Nsim}')
    Xsim[k] = cp.x

plt.figure()
plt.plot(t,Xsim[:,3],'r-')
plt.grid()
plt.xlabel('Time [s]')
plt.ylabel(f'$d\\theta/dt$')
plt.show()


# if you use Jupyter notebook make sure to pass notebook_mode = True in set_animation_figure
